lib/phoenix_html/link.ex

```diff
@@ -78,6 +83,11 @@ defmodule Phoenix.HTML.Link do
 
   def link(text, opts) do
     {to, opts} = pop_required_option!(opts, :to, "expected non-nil value for :to in link/2")
+
+    if invalid_destination?(to) do
+      raise ArgumentError, "link/2 expects a valid URL or path"
```

Comment:
What about:

> unsupported scheme given to link/2. In case you want to link to an unknown or unsafe scheme, such as javascript, use a tuple: {:javascript, rest}

Reply:
Definitely a better error. What are your thoughts with regard to:

> I do think transforming [the value of "to"] w/in link could be nice, because that would allow us to clear the invalid URL instead of raising (which becomes a DoS as opposed to an XSS).

As-is the onus is still on the programmer to appropriately filter bad input. It's better than an active attack, but still not the friendliest. An additional benefit of transforming the "to" value is that we could properly handle `{:javascript, to}` within link.ex.

That said, I do also see the value of raising.

Reply:
And, if we don't want to transform the input here, where would we want to do it?

Reply:
We have four different input combinations:

  1. invalid scheme and unsafe
  2. invalid scheme and safe
  3. valid scheme and unsafe
  4. valid scheme and safe

Currently we only check if a string has a valid scheme if it is unsafe. But do we want to consider `{:safe, "javascript:..."}` as a valid scheme or do we want to force the user to also take action on that?

I would say `{:safe, "javascript:..."}` is still invalid because someone may have done this: `link "Bar", to: html_escape(conn.params["foo"])` and that does not imply they support javascript. And even though html_escape is called implicitly, users can still call it if they want to.

If that's the case, then the implementation should probably do this:

```
defp invalid_destination?({:safe, to}) do
  invalid_destination?(to)
end
```

Does it make sense?

Regarding raising or not, the issue comes when the user passes `"javascript:..."` in development and then it is unclear why it does not work. So if we want to discard, we would need to at least log, but there is very likely other parts of Phoenix that would raise on invalid input, so I am not sure if it is worth going in this direction.